File: schnorr/schnorr.py
import random
from Crypto.Util.number import *

def keygen(bits=256):
    """Generate public and private key."""
    p = getPrime(bits)
    g = getRandomRange(2, p-1)
    x = getRandomRange(2, p-2)  # private key
    y = pow(g, x, p)             # public key
    return (p, g, y), x

def prover_commitment(p, g):
    """Step 1: Prover sends a commitment."""
    r = getRandomRange(2, p-2)
    t = pow(g, r, p)
    return t, r

def verifier_challenge(p):
    """Step 2: Verifier sends a challenge."""
    return getRandomRange(1, p-1)

def prover_response(r, c, x, p):
    """Step 3: Prover sends a response."""
    s = (r + c * x) % (p-1)
    return s

def verifier_check(p, g, y, t, c, s):
    """Verifier checks the prover's response."""
    return pow(g, s, p) == (t * pow(y, c, p)) % p

def schnorr_protocol(bits=256):
    """Demonstrate the Schnorr protocol."""
    # Key generation
    params, x = keygen(bits)
    p, g, y = params

    # Step 1: Commitment
    t, r = prover_commitment(p, g)

    # Step 2: Challenge
    c = verifier_challenge(p)

    # Step 3: Response
    s = prover_response(r, c, x, p) 

    # Verification
    if verifier_check(p, g, y, t, c, s):
        print("Verification successful!")
    else:
        print("Verification failed!")

schnorr_protocol()

File: schnorr/schnorr.go
package schnorr

import (
	"crypto/rand"
	"fmt"
	"math/big"
)

func Keygen(bits int) (p, g, y, x *big.Int) {
	p, _ = rand.Prime(rand.Reader, bits)
	g = randomRange(big.NewInt(2), new(big.Int).Sub(p, big.NewInt(1)))
	x = randomRange(big.NewInt(2), new(big.Int).Sub(p, big.NewInt(2)))
	y = new(big.Int).Exp(g, x, p)
	return
}

func ProverCommitment(p, g *big.Int) (t, r *big.Int) {
	r = randomRange(big.NewInt(2), new(big.Int).Sub(p, big.NewInt(2)))
	t = new(big.Int).Exp(g, r, p)
	return
}

func VerifierChallenge(p *big.Int) *big.Int {
	return randomRange(big.NewInt(1), new(big.Int).Sub(p, big.NewInt(1)))
}

func ProverResponse(r, c, x, p *big.Int) *big.Int {
	xc := new(big.Int).Mul(c, x)
	s := new(big.Int).Add(r, xc)
	return s.Mod(s, new(big.Int).Sub(p, big.NewInt(1)))
}

func VerifierCheck(p, g, y, t, c, s *big.Int) bool {
	gs := new(big.Int).Exp(g, s, p)
	yc := new(big.Int).Exp(y, c, p)
	tc := new(big.Int).Mul(t, yc)
	return gs.Cmp(tc.Mod(tc, p)) == 0
}

func schnorrProtocol(bits int) {
	p, g, y, x := Keygen(bits)
	t, r := ProverCommitment(p, g)
	c := VerifierChallenge(p)
	s := ProverResponse(r, c, x, p)

	if VerifierCheck(p, g, y, t, c, s) {
		fmt.Println("Verification successful")
	} else {
		fmt.Println("Verification failed")
	}
}

func schnorrProtocolPredefined(bits int, predefinedX *big.Int) {
    // Assuming that the predefinedX is legitimate and lies between 2 and p-2
    p, g, y, _ := Keygen(bits)
    x := predefinedX
    y = new(big.Int).Exp(g, x, p)  // Compute y using predefined x
    
    t, r := ProverCommitment(p, g)
    c := VerifierChallenge(p)
    s := ProverResponse(r, c, x, p)

    if VerifierCheck(p, g, y, t, c, s) {
        fmt.Println("Verification successful")
    } else {
        fmt.Println("Verification failed")
    }
}

func randomRange(min, max *big.Int) *big.Int {
	rangeInt := new(big.Int).Sub(max, min)
	n, _ := rand.Int(rand.Reader, rangeInt)
	return n.Add(n, min)
}

func main() {
	schnorrProtocol(256)
}
